### DISCUSSION TITLE
Can I use Jena Reasoners in Fuseki through SPARQL Endpoint?

### DISCUSSION THREAD

**Discussion by chenkuncloud** [apache/jena#1297] [Q&A] [answered]
I want to use Jena Reasoners in Fuseki through SPARQL Endpoint, such as JAVA API [InfModel](https://jena.apache.org/documentation/ontology/index.html):
```
InfModel infModel = ModelFactory.createInfModel(reasoner, model);
......
```
But I didn't find a suitable method in RDFConnection.

Like Virtuoso, it provides [VirtGraph](http://docs.openlinksw.com/virtuoso/rdfnativestorageprovidersjena/#rdfnativestorageprovidersjenaexamples3):
```
VirtGraph graph = new VirtGraph("http://test.com", "jdbc:virtuoso://localhost:1111", "dba", "mysecret");
VirtModel model = new VirtModel(graph);
InfModel infModel = ModelFactory.createInfModel(reasoner, model);
......
```

I know Fuseki can use Reasoners through modifying config.ttl, but I want to use JAVA API because my rules are dynamic.

I wonder if I missed something about Jena? Can I use Jena Reasoners in Fuseki through SPARQL Endpoint?

Best Regards

**Comment by afs** [comment, marked as answer]
So you want to modify the rules in a running Fuseki?

Fuseki is the server so the modification will have to be java code running in Fuseki. This is possible but not simple.

This only applies to the "Fuseki main" form of the server; the war file form would need repacking to add code. war files, by design, are effectively sealed.

[Fuseki Modules](https://jena.apache.org/documentation/fuseki2/fuseki-modules) is the way to add code into a Fuseki server. It can modify the server being built to add code and additional servlets for interacting with the server.

Or you can run a java program that builds and runs a Fuseki server. See `FusekiServer.create()` for the builder of servers that allows adding servlets/servlet filters and custom data services.

The other way is missing from Fuseki - the ability to cause a running server to reload its configuration and seamlessly switchover.

**Reply by chenkuncloud** [reply]
Thank you, I've learned that I have to modify the rules through `Fuseki server`.

I found method `configuration` after I `implements FusekiModule`, but method `configuration` is called just before the server is built.

I do not understand how to modify the rules. Can you give me an example of the code?

Best Regards

**Reply by chenkuncloud** [reply]
I also know that I can do it by [FusekiServer.create()](https://jena.apache.org/documentation/fuseki2/fuseki-embedded.html). But I still didn't find a way to modify the rules after `server.start()`.
```
FusekiServer server = FusekiServer.create()
	.parseConfigFile("config.ttl")
	.build();
server.start();
// How to modify the rules?
```

**Reply by afs** [reply]
Something like:

The dataset can be found in the `DataAccessPointRegistry` for the server.

Find the `InfGraph` in that dataset. Cast to `FBRuleInfGraph`, change the rules and call `rebind`.

You can more likely to get answers on the jena user mailing list.

**Comment by chenkuncloud** [comment]
Thank you very much for your reply !

This is what I want. I can start Fuseki through FusekiServer and modify the rules.
